<div style="text-align: center; padding: 20px; font-family: Vazir;">
<h1 align="center" style="font-size: 28px; color:rgb(64, 244, 202); width: 100%;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>تمرین ۲<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1>
<h2 style="color:rgb(90, 255, 184); font-size: 20px;">Logistic Regression and Naive Bayes</h2>
<p align="center" style="color: #666; font-size: 16px;">فرهاد نصری - علیرضا زمانی</p>
<p align="center" style="color: #666; font-size: 16px; margin-bottom: 30px;">farhadnasri999@gmail.com - shigzv@gmail.com</p>

<div dir="rtl" style="border: 2px dashed rgb(90, 255, 184); border-radius: 8px; padding: 20px; margin: 20px auto; max-width: 500px; text-align: right;">
<p style="color: rgb(64, 244, 202); font-size: 18px; margin-bottom: 15px;">📝 مشخصات دانشجو:</p>
<p style="color: #666; margin: 5px;">نام و نام خانوادگی: نوید ادیب</p>
<p style="color: #666; margin: 5px;">شماره دانشجویی: ۸۱۰۱۰۴۰۰۴</p>
<p style="color: #666; margin: 5px;">تاریخ ارسال: {{تاریخ_ارسال}}</p>
</div>
</div>

<div dir="rtl" style="text-align: justify; padding: 25px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<div style="padding-right:100px">
📋 <b>ساختار تمرین:</b>
<li><b>سوال اول - <span dir="ltr">Logistic Regression and Naive Bayes (from scratch)</span> (60)</b></li>
<ul>
<li>بخش اول: پیش پردازش داده‌های متنی و استفاده از Bag of Words</li>
<li>بخش دوم: جداسازی داده‌های آموزش و آزمایش</li>
<li>بخش سوم: پیاده‌سازی معیار‌های ارزیابی</li>
<li>بخش چهارم: پیاده‌سازی Logistic Regression</li>
<li>بخش پنجم: پیاده‌سازی Naive Bayes</li>
<li>بخش ششم: تحلیل نتایج</li>
</ul>
<li><b>سوال دوم - <span dir="ltr">Logistic Regression and Naive Bayes (e.g. 
sklearn)</span> (40)</b></li>
<ul>
<li>بخش اول: استخراج ویژگی‌های ساختاری</li>
<li>بخش دوم: استخراج ویژگی‌های آماری و محتوایی</li>
<li>بخش سوم: استخراج ویژگی‌های دلخواه</li>
<li>بخش چهارم: ترکیب همه ویژگی‌های استخراج‌شده</li>
</ul>
</div>
</div>
<div dir='rtl' style="line-height: 1.8; font-family: Vazir; font-size: 16px; margin-top: 20px; background-color: #e8eaf6; padding: 15px; border-radius: 8px; color:black">
💡 <b>نکات مهم:</b>
<br>
در سوال دو مجاز به استفاده از کتابخانه‌های آماده برای مدل‌ها و متریک‌ها هستید ولی در سوال یک، همان‌طور که در متن سوال هم ذکر شده، باید از پایه پیاده‌سازی کنید.
</div>
</div>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال اول: پیاده‌سازی Logistic Regression و Naive Bayes <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این سوال، شما باید این دو مدل پایه‌ای برای طبقه‌بندی متون را از صفر (بدون استفاده از کتابخانه‌های آماده) پیاده‌سازی کنید. هم‌چنین نیاز است، پیش‌پردازش داده‌ها و ارزیابی عملکرد مدل‌ها را نیز انجام دهید.




## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: پیش پردازش داده‌های متنی و استفاده از Bag of Words</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در گام پیش‌پردازش، هدف ما آماده‌سازی داده‌های متنی برای استفاده در مدل‌های یادگیری ماشین است. از آن‌جا که مدل‌ها با داده‌های عددی کار می‌کنند، باید متن خام را به شکلی عددی تبدیل کنیم تا بتوانند الگوهای موجود در واژه‌ها را بیاموزند. در این بخش، قصد داریم با تبدیل متون به نمایش Bag of Words، هر ایمیل را بر اساس تعداد تکرار واژه‌هایش به یک بردار عددی تبدیل کنیم.
<br>
دیتاست emails_1.csv شامل دو ستون است:
<br>
text (متن ایمیل)
<br>
status (برچسب، مشخص‌کننده‌ی spam یا ham بودن ایمیل)
<br>
شما باید مراحل زیر را انجام دهید:
<br>
۱. ستون status را به مقادیر عددی تبدیل کنید (spam → 1 و ham → 0).
<br>
۲. متن‌ها را پاک‌سازی کنید: تمام کاراکترهای غیر الفبایی (اعداد، علائم و غیره) را حذف و همه‌ی حروف را کوچک (lowercase) کنید.
<br>
۳. با استفاده از روش Bag of Words، فراوانی کلمات را محاسبه کنید و فقط ۱۵ کلمه‌ی پرتکرار را نگه دارید.
<br>
💡 نکته: برای این بخش می‌توانید از CountVectorizer در کتابخانه‌ی sklearn استفاده کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
یک دیتافریم جدید که شامل ۱۵ ویژگی + ستون status است. (تمام ۱۰ ردیف این دیتافریم را نمایش دهید.)
</p>

In [24]:
# Step 1: Import required libraries
import pandas as pd
import re
from sklearn.feature_extraction.text import CountVectorizer

In [25]:
# Step 2: Load the dataset
df = pd.read_csv("datasets/q1/emails_1.csv")
# Step 3: Convert labels to numeric values (spam → 1, ham → 0)
df['status'] = df['status'].map({'spam': 1, 'ham': 0})
df.head(2)

text  status
0  🎉 Congratulations! You have won a FREE ticket ...       1
1  Reminder: The project meeting will be held tom...       0

In [26]:
# Step 4: Define a text cleaning function
def clean_text(text):
    # Remove all non-alphabetic characters
    text = re.sub(r'[^a-zA-Z]', ' ', text)
    # Convert text to lowercase
    text = text.lower()
    # Remove extra spaces
    text = re.sub(r'\s+', ' ', text).strip()
    return text

# Step 5: Apply text cleaning to all messages
df['clean_text'] = df['text'].apply(clean_text)
df.head(2)

text  status  \
0  🎉 Congratulations! You have won a FREE ticket ...       1   
1  Reminder: The project meeting will be held tom...       0   

                                          clean_text  
0  congratulations you have won a free ticket to ...  
1  reminder the project meeting will be held tomo...

In [27]:
# Step 6: Initialize CountVectorizer to keep only top 15 frequent words
vectorizer = CountVectorizer(max_features=15)

# Step 7: Fit and transform the clean text into numerical vectors
X = vectorizer.fit_transform(df['clean_text']).toarray()

# Step 8: Retrieve the feature names
feature_names = vectorizer.get_feature_names_out()
print("Top 15 frequent words:",feature_names)

Top 15 frequent words: ['and' 'claim' 'enjoy' 'files' 'for' 'free' 'get' 'our' 'please' 'project'
 'the' 'ticket' 'to' 'you' 'your']


In [28]:
# Step 9: Compute total frequency of each word
word_counts = X.sum(axis=0)

# Step 10: Pair each word with its total count
word_freq = [(feature_names[i], word_counts[i]) for i in range(len(feature_names))]

# Step 11: Sort words by count (descending)
sorted_words = [word for word, count in sorted(word_freq, key=lambda x: x[1], reverse=True)]

# Step 12: Create the Bag of Words DataFrame using the sorted column order
bow_df = pd.DataFrame(X, columns=feature_names)
bow_df = bow_df[sorted_words]   # reorder columns

# Step 13: Add the target column (status) as the last column
bow_df['status'] = df['status'].values

# Step 14: Display the first few rows of the final DataFrame
print(bow_df.head(10))

   your  free  for  the  to  you  and  claim  get  our  please  project  \
0     1     2    0    0   1    1    0      1    0    1       0        0   
1     1     0    0    1   0    0    1      0    0    0       1        2   
2     0     2    2    0   0    1    0      0    1    1       0        0   
3     1     0    0    1   0    0    1      0    0    0       1        0   
4     2     1    1    0   0    1    0      1    0    0       0        0   
5     0     1    0    0   0    0    0      0    1    1       0        0   
6     0     0    1    1   0    0    0      0    0    0       1        1   
7     2     1    0    1   1    0    0      1    0    0       0        0   
8     1     0    1    1   1    1    1      0    1    0       0        0   
9     2     0    1    0   1    0    0      0    0    0       0        0   

   enjoy  files  ticket  status  
0      0      0       2       1  
1      0      1       0       0  
2      1      0       0       1  
3      0      0       0       0  
4   

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
از این مرحله به بعد، دیگر با فایل emails_1.csv و دیتافریم حاصل از آن کاری نخواهیم داشت. برای سهولت کار، فایل emails_2.csv در اختیار شما قرار گرفته است. این فایل، نتیجه‌ی همان فرایند پیش‌پردازش بر روی ۵۱۷۲ ایمیل واقعی است. به بیان ساده‌تر، در این مجموعه داده، هر ردیف نمایانگر یک ایمیل و هر ستون نشان‌دهنده‌ی فراوانی یکی از ۳۰۰۰ کلمه‌ی پرتکرار در کل داده‌ها است.
در ادامه، مدل‌هایی که پیاده‌سازی می‌کنید باید روی این مجموعه داده، آموزش داده شوند. بنابراین، در این بخش فایل emails_2.csv را بخوانید و محتوای آن را نمایش دهید تا با ساختار داده آشنا شوید.
</p>



<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
پنج ردیف اول دیتافریم مذکور را نمایش دهید.
</p>

In [29]:
import pandas as pd

In [30]:
# Load the dataset
df = pd.read_csv("datasets/q1/emails_2.csv")
# Display first 5 rows
print(df.head())
# Remove the non-feature column "Email No."
df = df.drop(columns=["Email No."])
# Check label distribution in the "Prediction" column
label_counts = df["Prediction"].value_counts()
print("\nLabel distribution:")
print(label_counts)

  Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0   Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1   Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2   Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3   Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4   Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   

   valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0       0    0               0         0         0   0    0           0  
1       0    0               0         0         0   1    0           0  
2       0    0               0         0         0   0    0           0  
3       0    0               0         0         0   0    0           0  
4       0    0               0         0         0   1    0           0  

[5 rows x 3002 columns]

Label distribution:
Prediction
0    3672
1    1500
Name: coun

<div style="direction: rtl; text-align: right; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">

### 📄 Dataset Summary

- The dataset contains **5172 emails** with **3002 columns**.  
- After removing the non-feature column **`Email No.`**, we have:
  - **3000 word-count features**
  - **1 label column (`Prediction`)**

---

### 🔍 Label Distribution

| Label | Meaning | Count |
|-------|---------|--------|
| **0** | Ham     | **3672** |
| **1** | Spam    | **1500** |

- The dataset is **moderately imbalanced**, with ~71% ham and ~29% spam.
</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: جداسازی داده‌های آموزش و آزمایش</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
داده‌ها را با نسبت ۸۰ درصد برای آموزش و ۲۰ درصد برای آزمایش تقسیم کنید. در انتها تعداد نمونه‌ها را چاپ کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
X_train, X_test, y_train, y_test<br>
تعداد نمونه‌های X_train و X_test 
</p>

In [31]:
import numpy as np

def stratified_train_test_split(X, y, test_size=0.2, seed=42):
    np.random.seed(seed)

    # Get indices for each class
    idx_ham  = np.where(y == 0)[0]
    idx_spam = np.where(y == 1)[0]

    # Shuffle
    np.random.shuffle(idx_ham)
    np.random.shuffle(idx_spam)

    # Test set sizes (stratified)
    n_test_ham  = int(len(idx_ham) * test_size)
    n_test_spam = int(len(idx_spam) * test_size)

    # Split ham
    test_idx_ham  = idx_ham[:n_test_ham]
    train_idx_ham = idx_ham[n_test_ham:]

    # Split spam
    test_idx_spam  = idx_spam[:n_test_spam]
    train_idx_spam = idx_spam[n_test_spam:]

    # Combine train/test indices
    train_indices = np.concatenate([train_idx_ham, train_idx_spam])
    test_indices  = np.concatenate([test_idx_ham,  test_idx_spam])

    # Shuffle final sets
    np.random.shuffle(train_indices)
    np.random.shuffle(test_indices)

    # Return in the order you want
    X_train = X[train_indices]
    X_test  = X[test_indices]
    y_train = y[train_indices]
    y_test  = y[test_indices]

    return X_train, X_test, y_train, y_test


In [32]:
X = df.drop(columns=["Prediction"]).values   # convert to numpy
y = df["Prediction"].values
X_train, X_test, y_train, y_test = stratified_train_test_split(X, y, test_size=0.2)

# Show results
print("Train size:", len(y_train))
print("Test size:", len(y_test))

print("\nTrain distribution:")
print(pd.DataFrame(y_train).value_counts())

print("\nTest distribution:")
print(pd.DataFrame(y_test).value_counts())

Train size: 4138
Test size: 1034

Train distribution:
0    2938
1    1200
Name: count, dtype: int64

Test distribution:
0    734
1    300
Name: count, dtype: int64


<div style="direction: ltr; text-align: left; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">
# Train-Test Split and Label Distribution

The dataset contains **imbalanced labels** (ham vs spam).  
To ensure that both training and test sets reflect the original class distribution, we used a **stratified split** with **80% for training** and **20% for testing**.

- **Train size:** 4138  
- **Test size:** 1034  

**Train distribution:**  
```
0 (ham): 2938  
1 (spam): 1200
```

**Test distribution:**  
```
0 (ham): 734  
1 (spam): 300
```

This stratification helps models learn effectively from both classes and ensures that evaluation metrics are reliable.


</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: پیاده‌سازی معیار‌های ارزیابی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
پیش از آن‌که وارد مرحله‌ی پیاده‌سازی و آموزش مدل‌های طبقه‌بندی شویم، لازم است معیارهایی برای سنجش عملکرد آن‌ها تعریف کنیم. این معیارها به ما کمک می‌کنند تا بفهمیم مدل تا چه اندازه در تشخیص درست نمونه‌های مثبت و منفی موفق عمل کرده است.
<br>
در این بخش، توابع زیر را برای حالت دودویی (binary classification) پیاده‌سازی کنید:
<br>
accuracy(y_true, y_pred) — نسبت پیش‌بینی‌های درست به کل نمونه‌ها
<br>
precision(y_true, y_pred) — درصد پیش‌بینی‌های مثبت که واقعاً مثبت بوده‌اند
<br>
recall(y_true, y_pred) — درصد نمونه‌های مثبت واقعی که مدل آن‌ها را درست شناسایی کرده است
<br>
f1_score(y_true, y_pred) — میانگین هارمونیک بین precision و recall، برای ایجاد توازن میان آن دو
<br>
هر تابع باید مقدار عددی متناظر با معیار مورد نظر را برگرداند.
<br>
 به توابع بالا دو ورودی زیر را بدهید و خروجی بگیرید:    y_true = [0, 1, 1, 0, 1] ---- y_pred = [0, 1, 0, 0, 1]
<br>
💡 نکته: برای این بخش می‌توانید از  کتابخانه‌ی numpy استفاده کنید.
</p>

In [33]:
# implementation of metrics
import numpy as np

# Calculate accuracy
def accuracy(y_true, y_pred):
    # Proportion of correct predictions
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)
    return np.mean(y_true == y_pred)

# Calculate precision
def precision(y_true, y_pred):
    # TP = predicted 1 and true 1
    # FP = predicted 1 but true 0
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)
    
    tp = np.sum((y_pred == 1) & (y_true == 1))
    fp = np.sum((y_pred == 1) & (y_true == 0))
    
    if tp + fp == 0:
        return 0.0
    return tp / (tp + fp)

# Calculate recall
def recall(y_true, y_pred):
    # TP = predicted 1 and true 1
    # FN = predicted 0 but true 1
    y_true = np.array(y_true)
    y_pred = np.array(y_pred)
    
    tp = np.sum((y_pred == 1) & (y_true == 1))
    fn = np.sum((y_pred == 0) & (y_true == 1))
    
    if tp + fn == 0:
        return 0.0
    return tp / (tp + fn)

# Calculate F1 score
def f1_score(y_true, y_pred):
    # Harmonic mean of precision and recall
    p = precision(y_true, y_pred)
    r = recall(y_true, y_pred)
    
    if p + r == 0:
        return 0.0
    return 2 * (p * r) / (p + r)

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
مثال:
<br>
y_true = [0, 1, 1, 0, 1]
<br>
y_pred = [0, 1, 0, 0, 1]
<br>
print(accuracy(y_true, y_pred))  # 0.80
<br>
print(precision(y_true, y_pred)) # 1.00
<br>
print(recall(y_true, y_pred))    # 0.66
<br>
print(f1_score(y_true, y_pred))  # 0.80
<br>
</p>

In [34]:
y_true = [0, 1, 1, 0, 1]
y_pred = [0, 1, 0, 0, 1]
print("Accuracy:", accuracy(y_true, y_pred))
print("Precision:", precision(y_true, y_pred))
print("Recall:", recall(y_true, y_pred))
print("F1 Score:", f1_score(y_true, y_pred))


Accuracy: 0.8
Precision: 1.0
Recall: 0.6666666666666666
F1 Score: 0.8


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: پیاده‌سازی Logistic Regression</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
کلاس Logistic Regression را به طور کامل پیاده سازی کنید.سپس مدل را بر روی داده‌های train آموزش دهید و در انتها دقت مدل را باتوجه به متریک های تعریف شده بر روی داده‌های test گزارش کنید.
<br>
💡استفاده از هایپر پارامتر‌های مناسب مانند نرخ یادگیری و تعداد دوره‌های آموزش بر عهده خودتان است.
<br>
💡میتوانید قبل از آموزش داده ها را نرمال سازی کنید.
<br>
💡مجاز به استفاده از کتابخانه اماده نیستید.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
معیار‌های پیاده سازی شده را بر روی داده‌های test خروجی بگیرید و نتایج را چاپ کنید
</p>

In [35]:
import numpy as np


class LogisticRegressionScratch:
    def __init__(self, lr=0.01, epochs=100, batch_size=32, normalize=True):
        # Learning rate
        self.lr = lr
        # Number of training epochs
        self.epochs = epochs
        # Mini-batch size
        self.batch_size = batch_size
        # Whether to apply normalization
        self.normalize = normalize
        
        # Will be set during fitting
        self.W = None
        self.b = None
        self.mean = None
        self.std = None

    # --------------------------------------------
    # Sigmoid function
    # --------------------------------------------
    def sigmoid(self, z):
        return 1 / (1 + np.exp(-z))

    # --------------------------------------------
    # Binary cross-entropy loss
    # --------------------------------------------
    def compute_loss(self, y_true, y_pred):
        # Clip values for numerical stability
        y_pred = np.clip(y_pred, 1e-10, 1 - 1e-10)
        return -np.mean(y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred))

    # --------------------------------------------
    # Forward pass
    # --------------------------------------------
    def forward(self, X):
        return self.sigmoid(np.dot(X, self.W) + self.b)

    # --------------------------------------------
    # Normalize features
    # --------------------------------------------
    def normalize_features(self, X):
        self.mean = X.mean(axis=0)
        self.std = X.std(axis=0) + 1e-10  # avoid division by zero
        return (X - self.mean) / self.std

    # --------------------------------------------
    # Apply stored normalization parameters
    # --------------------------------------------
    def apply_normalization(self, X):
        return (X - self.mean) / self.std

    # --------------------------------------------
    # Training loop (SGD)
    # --------------------------------------------
    def fit(self, X, y):
        X = np.array(X)
        y = np.array(y)

        # Normalize if enabled
        if self.normalize:
            X = self.normalize_features(X)

        n_samples, n_features = X.shape

        # Initialize weights and bias
        self.W = np.zeros(n_features)
        self.b = 0.0

        # Determine progress checkpoints
        checkpoints = {int(self.epochs * p / 100) for p in range(10, 101, 10)}

        for epoch in range(1, self.epochs + 1):

            # Shuffle the data
            indices = np.random.permutation(n_samples)
            X = X[indices]
            y = y[indices]

            # Mini-batch SGD
            for start in range(0, n_samples, self.batch_size):
                end = start + self.batch_size
                X_batch = X[start:end]
                y_batch = y[start:end]

                # Forward pass
                y_pred = self.forward(X_batch)

                # Gradient calculation
                error = y_pred - y_batch
                dW = np.dot(X_batch.T, error) / len(X_batch)
                db = np.mean(error)

                # Gradient descent update
                self.W -= self.lr * dW
                self.b -= self.lr * db

            # Print progress at each 10%
            if epoch in checkpoints:
                percent = int(epoch / self.epochs * 100)
                print(f"Training progress: {percent}%")

        return self

    # --------------------------------------------
    # Predict probability
    # --------------------------------------------
    def predict_proba(self, X):
        X = np.array(X)

        if self.normalize:
            X = self.apply_normalization(X)

        return self.forward(X)

    # --------------------------------------------
    # Predict class (0/1)
    # --------------------------------------------
    def predict(self, X):
        proba = self.predict_proba(X)
        return (proba >= 0.5).astype(int)


In [36]:
model = LogisticRegressionScratch(
    lr=0.05,
    epochs=20,
    batch_size=64,
    normalize=True
)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

Training progress: 10%
Training progress: 20%
Training progress: 30%
Training progress: 40%
Training progress: 50%
Training progress: 60%
Training progress: 70%
Training progress: 80%
Training progress: 90%
Training progress: 100%


In [37]:
print("Accuracy:", accuracy(y_test, y_pred))
print("Precision:", precision(y_test, y_pred))
print("Recall:", recall(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

Accuracy: 0.965183752417795
Precision: 0.9230769230769231
Recall: 0.96
F1 Score: 0.9411764705882353


<div style="direction: ltr; text-align: left; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">


# Logistic Regression (Summary)

**Logistic Regression** is a supervised **binary classification** algorithm. It models the probability that an input belongs to the positive class.

## Model
The model computes:

$$
z = W^T X + b
$$

and applies the **sigmoid** function:

$$
\hat{y} = \sigma(z) = \frac{1}{1 + e^{-z}}
$$

which outputs a probability between **0 and 1**.

## Loss Function
Training uses **Binary Cross-Entropy**:

$$
L = -\frac{1}{m} \sum_{i=1}^{m} \Big[ y_i \log(\hat{y}_i) + (1-y_i) \log(1-\hat{y}_i) \Big]
$$

## Optimization
Parameters are updated using **gradient descent**:

$$
W := W - \alpha \frac{\partial L}{\partial W}, \quad
b := b - \alpha \frac{\partial L}{\partial b}
$$

## Prediction
Predict **1** if 

$$
\hat{y} \ge 0.5
$$

otherwise **0**.

## Why Use It?
- Simple and fast  
- Interpretable  
- Produces calibrated probabilities

</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش پنجم: پیاده‌سازی Naive Bayes</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
کلاس Multinomial Naive Bayes را به طور کامل پیاده سازی کنید. سپس مدل را بر روی داده‌های train آموزش دهید و در انتها دقت مدل را باتوجه به معیار های تعریف شده بر روی داده‌های test گزارش کنید.
<br>
- چرا از مدل‌های دیگر مانند Gaussian Naive Bayes استفاده نکردیم؟ آیا از مدل‌های دیگر Naive Bayes می‌توان استفاده کرد؟
<br>
💡مجاز به استفاده از کتابخانه اماده نیستید.
<br>
💡برای آشنایی با Naive Bayes می‌توانید به Appendix کتاب jurafsky مراجعه کنید.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
معیار‌های پیاده سازی شده را بر روی داده‌های test خروجی بگیرید و نتایج را چاپ کنید
</p>

In [22]:
import numpy as np

class MultinomialNaiveBayes:
    def __init__(self, alpha=1.0, normalize=False):
        """
        alpha: Laplace smoothing
        normalize: if True, normalize features before training (not needed for MNB)
        """
        self.alpha = alpha
        self.normalize = normalize
        
        self.class_log_priors = None
        self.feature_log_probs = None
        self.mean = None
        self.std = None
        
    # --------------------------------------------
    # Normalize features (optional, rarely needed)
    # --------------------------------------------
    def normalize_features(self, X):
        self.mean = X.mean(axis=0)
        self.std = X.std(axis=0) + 1e-10
        return (X - self.mean) / self.std
    
    def apply_normalization(self, X):
        return (X - self.mean) / self.std

    # --------------------------------------------
    # Fit the Multinomial Naive Bayes model
    # --------------------------------------------
    def fit(self, X, y, epochs=1):
        """
        Since NB is a closed-form algorithm, epochs=1 is enough.
        But added epochs only for 10% progress reporting.
        """
        X = np.array(X)
        y = np.array(y)
        
        # Optional normalization
        if self.normalize:
            X = self.normalize_features(X)

        n_samples, n_features = X.shape
        classes = np.unique(y)
        n_classes = len(classes)

        # Initialize matrices
        class_counts = np.zeros(n_classes)
        feature_counts = np.zeros((n_classes, n_features))

        # Progress checkpoints (10,20,...,100)
        checkpoints = {int(epochs * p / 100) for p in range(10, 101, 10)}

        for epoch in range(1, epochs + 1):

            # Counting features and class frequencies
            for idx, cls in enumerate(classes):
                X_c = X[y == cls]
                class_counts[idx] = X_c.shape[0]
                feature_counts[idx] = X_c.sum(axis=0)

            # Compute class prior: P(y)
            self.class_log_priors = np.log(class_counts / n_samples)

            # Compute conditional probabilities with Laplace smoothing
            # P(word_i | class)
            smoothed_feature_counts = feature_counts + self.alpha
            smoothed_totals = smoothed_feature_counts.sum(axis=1).reshape(-1, 1)
            self.feature_log_probs = np.log(smoothed_feature_counts / smoothed_totals)

            # Print progress
            if epoch in checkpoints:
                print(f"Training progress: {int(epoch / epochs * 100)}%")

        return self

    # --------------------------------------------
    # Predict log-probabilities for each class
    # --------------------------------------------
    def predict_log_proba(self, X):
        X = np.array(X)
        if self.normalize:
            X = self.apply_normalization(X)

        # log P(y) + sum (count_i * log P(word_i | y))
        return self.class_log_priors + np.dot(X, self.feature_log_probs.T)

    # --------------------------------------------
    # Predict class labels
    # --------------------------------------------
    def predict(self, X):
        log_probs = self.predict_log_proba(X)
        return np.argmax(log_probs, axis=1)


In [27]:
model = MultinomialNaiveBayes(alpha=1)
model.fit(X_train, y_train, epochs=10)   # epochs only for showing progress
y_pred = model.predict(X_test)

Training progress: 10%
Training progress: 20%
Training progress: 30%


Training progress: 40%
Training progress: 50%
Training progress: 60%
Training progress: 70%
Training progress: 80%
Training progress: 90%
Training progress: 100%


In [28]:
print("Accuracy:", accuracy(y_test, y_pred))
print("Precision:", precision(y_test, y_pred))
print("Recall:", recall(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

Accuracy: 0.937137330754352
Precision: 0.8637770897832817
Recall: 0.93
F1 Score: 0.8956661316211878


<div style="direction: ltr; text-align: left; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">

# Multinomial Naive Bayes (Summary)

**Multinomial Naive Bayes (MNB)** is a supervised **probabilistic classifier** commonly used for **text classification** tasks like spam detection and sentiment analysis.

## Model
MNB models the probability of a class given a document using:

$$
P(y \mid X) \propto P(y) \prod_{i=1}^{n} P(w_i \mid y)^{x_i}
$$

Where:

- \(P(y)\) = prior probability of class \(y\)  
- \(P(w_i \mid y)\) = probability of word \(i\) given class \(y\)  
- \(x_i\) = count of word \(i\) in the document  

The **log probabilities** are often used to avoid numerical underflow:

$$
\log P(y \mid X) = \log P(y) + \sum_{i=1}^{n} x_i \log P(w_i \mid y)
$$

## Training
- Count occurrences of each word per class  
- Apply **Laplace smoothing** to avoid zero probabilities:

$$
P(w_i \mid y) = \frac{\text{count}(w_i, y) + \alpha}{\sum_j \text{count}(w_j, y) + \alpha V}
$$

Where \(\alpha\) is usually 1 and \(V\) is vocabulary size.

## Prediction
Predict the class with the **highest posterior probability**:

$$
\hat{y} = \arg\max_y \log P(y \mid X)
$$

## Why Multinomial NB?
- Works well with **count-based features** (Bag-of-Words, TF-IDF)  
- Very fast and simple  
- Handles sparse high-dimensional data efficiently  

## Why Not Gaussian NB?
- Gaussian NB assumes **continuous features with Gaussian distribution**  
- Text features are **counts (integers)** and non-negative  
- Using Gaussian NB can produce incorrect probabilities and poor performance  

---


</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;"> بخش ششم: تحلیل نتایج</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نتایج دو مدل پیاده سازی شده را با یکدیگر مقایسه کنید و تحلیلی بر نتایج داشته باشید.<br>
    
</p>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال دوم: استفاده از Logistic Regression و Naive Bayes <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">هدف این سوال، بررسی و مقایسه عملکرد مدل‌های Logistic Regression و Naive Bayes در تشخیص لینک‌های فیشینگ از لینک‌های قانونی (معتبر) است.
شما باید با استفاده از مجموعه‌داده‌ی ارائه‌شده (آدرس‌های اینترنتی خام)، یک سری ویژگی‌های عددی از این URLها استخراج نمایید و تأثیر انتخاب ویژگی‌ها و نرمال‌سازی داده‌ها را بر عملکرد مدل‌ها بررسی کنید.<br>نکته: در این سوال مجاز به استفاده از کتاب‌خانه‌های آماده هستید.


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: ویژگی‌های ساختاری</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی زیر را از آدرس استخراج کنید:
<br>
- تعداد نقطه‌ها (nb_dots)
<br>
- تعداد اسلش‌ها (nb_slashes)
<br>
- تعداد خط ‌تیره‌ها (nb_hyphens)
<br>
سپس با این سه ویژگی مدل‌ها را روی هشتاد درصد داده‌ها (داده‌های آموزش) آموزش دهید و روی بیست درصد داده‌ها (داده‌های آزمون) مقایسه کنید
<br>
<br>
- آیا لازم است این ویژگی‌ها را نرمال‌سازی کنید؟ چرا؟ اگر پاسخ شما «بله» است، نوع نرمال‌سازی را توضیح دهید و آن را اجرا کنید.
<br>
- به نظر شما از بین نسخه‌های مختلف Naive Bayes (GaussianNB یا MultinomialNB) کدام برای داده‌های شما مناسب‌تر است؟ دلیل خود را بنویسید و با آن مدل آزمایش را انجام دهید.
</p>

In [50]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB,MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score


In [42]:
# Load data
df = pd.read_csv("datasets/q2/urls.csv")
df["status"].value_counts()
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11430 entries, 0 to 11429
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   url     11430 non-null  object
 1   status  11430 non-null  object
dtypes: object(2)
memory usage: 178.7+ KB


In [43]:
# Convert labels to numeric: legitimate→0, phishing→1
df['label'] = df['status'].map({'legitimate': 0, 'phishing': 1})

# Extract features
df['nb_dots'] = df['url'].apply(lambda x: x.count('.'))
df['nb_slashes'] = df['url'].apply(lambda x: x.count('/'))
df['nb_hyphens'] = df['url'].apply(lambda x: x.count('-'))

X = df[['nb_dots', 'nb_slashes', 'nb_hyphens']]
y = df['label']
df[['nb_dots', 'nb_slashes', 'nb_hyphens','label']].sample(2)
print("label balance check")
df['label'].value_counts()

label balance check


label
0    5715
1    5715
Name: count, dtype: int64

In [48]:
# Stratified train-test split (80/20)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)
print(f"Num of Train samples: {len(y_train)}")
print(y_train.value_counts())
print(f"\nNum of Test samples: {len(y_test)}")
print(y_test.value_counts())

Num of Train samples: 9144
label
0    4572
1    4572
Name: count, dtype: int64

Num of Test samples: 2286
label
1    1143
0    1143
Name: count, dtype: int64


In [51]:
# ============================
# Logistic Regression
# (Needs Scaling)
# ============================
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

lr = LogisticRegression()
lr.fit(X_train_scaled, y_train)
y_pred_lr = lr.predict(X_test_scaled)

# ============================
# GaussianNB
# ============================
gnb = GaussianNB()
gnb.fit(X_train, y_train)
y_pred_gnb = gnb.predict(X_test)

# ============================
# 6. MultinomialNB
# ============================
mnb = MultinomialNB()
mnb.fit(X_train, y_train)
y_pred_mnb = mnb.predict(X_test)

<div style="direction: ltr; text-align: left; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">
# URL Phishing Detection – Feature Engineering and Model Evaluation

## 1. Extracted Features
From each URL, the following three numerical features were extracted:

- **nb_dots** – number of "." characters  
- **nb_slashes** – number of "/" characters  
- **nb_hyphens** – number of "-" characters  

These structural URL-based features are simple but useful for distinguishing phishing from legitimate URLs.

---

## 2. Train–Test Split
The dataset was divided using an **80/20 stratified split** to preserve the original class balance of:

- 0 → legitimate  
- 1 → phishing  

Stratification ensures the same ratio of classes in both train and test sets.

---

## 3. Normalization – Is It Necessary?
Yes — **normalization is necessary for Logistic Regression**, because it is a distance-based linear model and is sensitive to the scale of features.

### Applied method:
- **StandardScaler (Z-score normalization)**  
  $$
  x_{norm} = \frac{x - \mu}{\sigma}
  $$

### For Naive Bayes:
- **GaussianNB**: does *not* require scaling  
- **MultinomialNB**: should *not* be scaled (requires non-negative discrete counts)

Therefore:
- Logistic Regression → **scaled**  
- GaussianNB → **not scaled**  
- MultinomialNB → **not scaled**

---

## 4. Which Naive Bayes Variant Is Suitable?
Between **GaussianNB** and **MultinomialNB**:

### ✔ GaussianNB is theoretically the correct choice  
Because:
- our features (`nb_dots`, `nb_slashes`, `nb_hyphens`) are **continuous numeric counts**,  
- GaussianNB assumes **continuous features drawn from a normal distribution**,  
- MultinomialNB assumes **discrete word counts from text**, not general numeric features.

### However:
Both models were tested experimentally.

---

## 5. Model Results

| Model                | Accuracy | Precision | Recall | F1 Score |
|----------------------|----------|-----------|--------|----------|
| **Logistic Regression** | **0.6763** | 0.7042 | 0.6080 | **0.6526** |
| **Multinomial NB**      | 0.5424 | 0.5286 | **0.7830** | 0.6312 |
| **Gaussian NB**         | **0.6824** | **0.8243** | 0.4637 | 0.5935 |

---

## 6. Interpretation

### Logistic Regression
- Balanced performance.  
- Highest **F1 score**, making it the most reliable overall.  
- Scaling improved stability.

### Multinomial NB
- Lowest accuracy, but **highest recall** → detects more phishing URLs.  
- Useful when false negatives (missed phishing) are more dangerous.

### Gaussian NB
- Highest precision, good accuracy, but low recall.  
- More conservative → often predicts “legitimate”.  
- Matches the nature of our numeric features but suffers from small dataset size.

---

## 7. Final Conclusion

- **Best overall model → Logistic Regression**  
  It provides the strongest balance between detecting phishing and avoiding false alarms.

- **Best for maximum phishing detection → Multinomial NB**  
  Highest recall.

- **Best for minimizing false positives → Gaussian NB**  
  Highest precision.

GaussianNB is the *theoretically correct* variant for this dataset,  
but Logistic Regression delivered the **best practical performance**.

</div>


<div style="direction: ltr; text-align: left; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">
# URL Phishing Detection – Final Model Comparison

After extracting the numerical URL features (`nb_dots`, `nb_slashes`, `nb_hyphens`) and performing an 80/20 stratified split, three models were trained:

- **Logistic Regression**
- **Multinomial Naive Bayes**
- **Gaussian Naive Bayes**

The following evaluation metrics were computed on the test set:

---

## 📊 Model Performance

| Model               | Accuracy | Precision | Recall | F1 Score |
|---------------------|----------|-----------|--------|----------|
| **Logistic Regression** | **0.6763** | 0.7042 | 0.6080 | 0.6526 |
| **Multinomial NB**      | 0.5424 | 0.5286 | **0.7830** | 0.6312 |
| **Gaussian NB**         | **0.6824** | **0.8243** | 0.4637 | 0.5935 |

---

## 📌 Interpretation

### Logistic Regression
- Balanced performance.
- Good precision and decent recall.
- Strong overall F1 score → best compromise between false positives and false negatives.

### Multinomial NB
- Lowest accuracy, but **highest recall**, meaning it catches more phishing samples.
- Useful when missing a phishing URL is more dangerous than misclassifying a legitimate one.

### Gaussian NB
- Highest accuracy and precision.
- But recall is low → tends to miss phishing URLs.
- Indicates the model is conservative and predicts “legitimate” more often.

---

## 🎯 Conclusion

- **Best overall model:** Logistic Regression  
  (highest F1 among balanced models)

- **Best for catching phishing URLs:** Multinomial NB  
  (highest recall)

- **Best for avoiding false alarms:** Gaussian NB  
  (highest precision)


</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [62]:
# ============================
# Evaluation Function
# ============================
def evaluate(name, y_true, y_pred):
    return {
        "Model": name,
        "Accuracy": round(accuracy_score(y_true, y_pred), 4),
        "Precision": round(precision_score(y_true, y_pred), 4),
        "Recall": round(recall_score(y_true, y_pred), 4),
        "F1 Score": round(f1_score(y_true, y_pred), 4)
    }

results = []
results.append(evaluate("Logistic Regression", y_test, y_pred_lr))
results.append(evaluate("Multinomial NB", y_test, y_pred_mnb))
results.append(evaluate("Gaussian NB", y_test, y_pred_gnb))

# ============================
# 8. Beautiful Result Table
# ============================
results_df = pd.DataFrame(results)
print(results_df.to_string(index=False))

              Model  Accuracy  Precision  Recall  F1 Score
Logistic Regression    0.6763     0.7042  0.6080    0.6526
     Multinomial NB    0.5424     0.5286  0.7830    0.6312
        Gaussian NB    0.6824     0.8243  0.4637    0.5935


In [60]:
X_temp = X.copy()
X_temp["status"]=y
X_temp.head(5)
# del X_temp

nb_dots  nb_slashes  nb_hyphens  status
0        3           3           0       0
1        1           5           0       1
2        4           5           1       1
3        2           2           0       0
4        2           5           2       0

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: ویژگی‌های آماری و محتوایی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی زیر را از آدرس استخراج کنید:
<br>
- طول کل آدرس (length_url)
<br>
- نسبت تعداد ارقام (0–9) به کل طول آدرس (ratio_digits_url)
<br>
- طولانی‌ترین کلمه (کاراکترهای الفبایی) در URL که با علائم جداساز (نقطه، علامت‌سوال، اسلش و...) از هم جدا شده‌اند. (longest_words_raw)
<br>
سپس با این سه ویژگی مدل‌ها را روی هشتاد درصد داده‌ها (داده‌های آموزش) آموزش دهید و روی داده‌های آزمون (بیست درصد داده‌ها) مقایسه کنید.
<br>
- به نظر شما استفاده از Naive Bayes برای این ویژگی‌ها منطقی است؟ توضیح دهید. کدام نسخه مناسب‌تر است؟ 
</p>

<p style="direction: rtl; text-align: right; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">
✍️ <b>پاسخ تشریحی بخش دوم:</b><br>
{{پاسخ_خود_را_اینجا_بنویسید}}
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [71]:
import pandas as pd
import re

# Load data
df = pd.read_csv("datasets/q2/urls.csv")

# Convert labels
df['label'] = df['status'].map({'legitimate':0, 'phishing':1})

# Feature 1: URL length (apply log1p)
df['length_url'] = df['url'].apply(len)
df['length_url_log'] = np.log1p(df['length_url'])

# Feature 2: ratio of digits
df['ratio_digits_url'] = df['url'].apply(lambda x: sum(c.isdigit() for c in x)/len(x))

# Feature 3: longest alphabetic word
def longest_word(url):
    words = re.split(r'[./?&=-]', url)
    words_alpha = [w for w in words if w.isalpha()]
    return max([len(w) for w in words_alpha], default=0)
df['longest_words'] = df['url'].apply(longest_word)
df['longest_words_log'] = np.log1p(df['longest_words'])


df[['length_url', 'ratio_digits_url', 'longest_words','label']].head()

length_url  ratio_digits_url  longest_words  label
0          37          0.000000             11      0
1          77          0.220779             19      1
2         126          0.150794             13      1
3          18          0.000000              5      0
4          55          0.000000             11      0

In [72]:
print("using log of length may help")
df[['length_url_log', 'ratio_digits_url', 'longest_words_log','label']].head()

using log of length may help


length_url_log  ratio_digits_url  longest_words_log  label
0        3.637586          0.000000           2.484907      0
1        4.356709          0.220779           2.995732      1
2        4.844187          0.150794           2.639057      1
3        2.944439          0.000000           1.791759      0
4        4.025352          0.000000           2.484907      0

In [73]:
# ----------------------------
# Define feature sets
# ----------------------------
no_log = ['length_url','ratio_digits_url','longest_words']
yes_log = ['length_url_log','ratio_digits_url','longest_words_log']

# ----------------------------
# Train/Test Split (stratified)
# ----------------------------
# No-log features
X_no_log = df[no_log]
y = df['label']
X_train_no_log, X_test_no_log, y_train, y_test = train_test_split(
    X_no_log, y, test_size=0.2, stratify=y, random_state=42
)

# Log features
X_log = df[yes_log]
X_train_log, X_test_log, _, _ = train_test_split(
    X_log, y, test_size=0.2, stratify=y, random_state=42
)

In [74]:
# ----------------------------
# Scaling for Logistic Regression
# ----------------------------
scaler_no_log = StandardScaler()
X_train_scaled_no_log = scaler_no_log.fit_transform(X_train_no_log)
X_test_scaled_no_log = scaler_no_log.transform(X_test_no_log)

scaler_log = StandardScaler()
X_train_scaled_log = scaler_log.fit_transform(X_train_log)
X_test_scaled_log = scaler_log.transform(X_test_log)

In [75]:
# ----------------------------
# Models
# ----------------------------
def evaluate_models(X_train_scaled, X_test_scaled, X_train_raw, X_test_raw, y_train, y_test):
    # Logistic Regression
    lr = LogisticRegression(max_iter=1000)
    lr.fit(X_train_scaled, y_train)
    y_pred_lr = lr.predict(X_test_scaled)

    # Gaussian NB
    gnb = GaussianNB()
    gnb.fit(X_train_raw, y_train)
    y_pred_gnb = gnb.predict(X_test_raw)

    # Multinomial NB (not recommended)
    mnb = MultinomialNB()
    mnb.fit(X_train_raw, y_train)
    y_pred_mnb = mnb.predict(X_test_raw)

    # Evaluation
    def eval_model(name, y_true, y_pred):
        return {
            "Model": name,
            "Accuracy": round(accuracy_score(y_true, y_pred),4),
            "Precision": round(precision_score(y_true, y_pred),4),
            "Recall": round(recall_score(y_true, y_pred),4),
            "F1 Score": round(f1_score(y_true, y_pred),4)
        }

    results = []
    results.append(eval_model("Logistic Regression", y_test, y_pred_lr))
    results.append(eval_model("Gaussian NB", y_test, y_pred_gnb))
    results.append(eval_model("Multinomial NB", y_test, y_pred_mnb))

    return pd.DataFrame(results)

# Evaluate no-log features
results_no_log = evaluate_models(
    X_train_scaled_no_log, X_test_scaled_no_log,
    X_train_no_log, X_test_no_log,
    y_train, y_test
)

# Evaluate log features
results_log = evaluate_models(
    X_train_scaled_log, X_test_scaled_log,
    X_train_log, X_test_log,
    y_train, y_test
)

# Display results
print("=== Results without log features ===")
print(results_no_log.to_string(index=False))
print("\n=== Results with log features ===")
print(results_log.to_string(index=False))

=== Results without log features ===
              Model  Accuracy  Precision  Recall  F1 Score
Logistic Regression    0.6500     0.6974  0.5302    0.6024
        Gaussian NB    0.6382     0.7565  0.4077    0.5298
     Multinomial NB    0.5875     0.6006  0.5223    0.5587

=== Results with log features ===
              Model  Accuracy  Precision  Recall  F1 Score
Logistic Regression    0.6452     0.6861  0.5354    0.6015
        Gaussian NB    0.6369     0.7220  0.4453    0.5509
     Multinomial NB    0.6479     0.7076  0.5039    0.5887


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: ویژگی‌های خلاقانه</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی به انتخاب خودتان از آدرس‌ها استخراج نمایید و مشابه دو بخش قبلی، مدل‌ها را آموزش دهید.
برای هر ویژگی توضیح دهید چرا ممکن است در تشخیص فیشینگ مؤثر باشد؟
</p>

<p style="direction: rtl; text-align: right; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">
✍️ <b>پاسخ تشریحی بخش سوم:</b><br>
{{پاسخ_خود_را_اینجا_بنویسید}}
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [8]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB,MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [10]:
import pandas as pd
import numpy as np
import re

# Load data
df = pd.read_csv("datasets/q2/urls.csv")

# Convert labels
df['label'] = df['status'].map({'legitimate':0, 'phishing':1})

# Features
df['has_https'] = df['url'].apply(lambda x: 1 if x.startswith("https://") else 0)
df['has_www'] = df['url'].apply(lambda x: 1 if "://www." in x else 0)

def length_after_last_slash(url):
    # remove trailing slash (to avoid empty result)
    url = url.rstrip('/')
    # split and take the last segment
    last_part = url.split('/')[-1]
    return  np.log1p(len(last_part))

df['length_after_last_slash'] = df['url'].apply(length_after_last_slash)
X = df[['has_https', 'has_www', 'length_after_last_slash']]
y = df['label']
df[['has_https', 'has_www', 'length_after_last_slash','label']].sample(2)

has_https  has_www  length_after_last_slash  label
8439          0        0                 2.302585      1
5444          0        1                 3.218876      0

In [11]:
# ----------------------------
# Train/Test Split (stratified)
# ----------------------------
y = df['label']
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)
# ----------------------------
# Scaling for Logistic Regression
# ----------------------------
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [13]:
# ----------------------------
# Models
# ----------------------------
def evaluate_models(X_train_scaled, X_test_scaled, X_train_raw, X_test_raw, y_train, y_test):
    # Logistic Regression
    lr = LogisticRegression(max_iter=1000)
    lr.fit(X_train_scaled, y_train)
    y_pred_lr = lr.predict(X_test_scaled)

    # Gaussian NB
    gnb = GaussianNB()
    gnb.fit(X_train_raw, y_train)
    y_pred_gnb = gnb.predict(X_test_raw)

    # Multinomial NB (not recommended)
    mnb = MultinomialNB()
    mnb.fit(X_train_raw, y_train)
    y_pred_mnb = mnb.predict(X_test_raw)

    # Evaluation
    def eval_model(name, y_true, y_pred):
        return {
            "Model": name,
            "Accuracy": round(accuracy_score(y_true, y_pred),4),
            "Precision": round(precision_score(y_true, y_pred),4),
            "Recall": round(recall_score(y_true, y_pred),4),
            "F1 Score": round(f1_score(y_true, y_pred),4)
        }

    results = []
    results.append(eval_model("Logistic Regression", y_test, y_pred_lr))
    results.append(eval_model("Gaussian NB", y_test, y_pred_gnb))
    results.append(eval_model("Multinomial NB", y_test, y_pred_mnb))

    return pd.DataFrame(results)

# Evaluate no-log features
results = evaluate_models(
    X_train_scaled, X_test_scaled,
    X_train, X_test,
    y_train, y_test
)
print("\n=== Results with log features ===")
print(results.to_string(index=False))


=== Results with log features ===
              Model  Accuracy  Precision  Recall  F1 Score
Logistic Regression    0.7297     0.7033  0.7944    0.7461
        Gaussian NB    0.7358     0.7016  0.8206    0.7565
     Multinomial NB    0.7297     0.7033  0.7944    0.7461


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: ترکیب همه ویژگی‌ها</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
تمام ۹ ویژگی (۳ ساختاری + ۳ آماری + ۳ خلاقانه) را با هم ترکیب کنید.
سپس هر دو مدل را دوباره آموزش دهید و نتایج را ارزیابی کنید.
<br>
آیا ترکیب همه ویژگی‌ها باعث بهبود عملکرد مدل می‌شود؟
<br>
اگر خیر، به نظر شما دلیل آن چیست؟ (مثلاً تداخل بین ویژگی‌ها یا همبستگی بالا و...)
</p>


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB,MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [ ]:
# Load data
df = pd.read_csv("datasets/q2/urls.csv")

# Convert labels
df['label'] = df['status'].map({'legitimate':0, 'phishing':1})

In [18]:
# Features (custom)
df['has_https'] = df['url'].apply(lambda x: 1 if x.startswith("https://") else 0)
df['has_www'] = df['url'].apply(lambda x: 1 if "://www." in x else 0)

def length_after_last_slash(url):
    # remove trailing slash (to avoid empty result)
    url = url.rstrip('/')
    # split and take the last segment
    last_part = url.split('/')[-1]
    return  np.log1p(len(last_part))

df['length_after_last_slash'] = df['url'].apply(length_after_last_slash)

# Features (structural)
df['nb_dots'] = df['url'].apply(lambda x: x.count('.'))
df['nb_slashes'] = df['url'].apply(lambda x: x.count('/'))
df['nb_hyphens'] = df['url'].apply(lambda x: x.count('-'))

# Features (statistical)
df['length_url'] = np.log1p(df['url'].apply(len))
df['ratio_digits_url'] = df['url'].apply(lambda x: sum(c.isdigit() for c in x)/len(x))
def longest_word(url):
    words = re.split(r'[./?&=-]', url)
    words_alpha = [w for w in words if w.isalpha()]
    return np.log1p(max([len(w) for w in words_alpha], default=0))
df['longest_words'] = df['url'].apply(longest_word)

X = df[["has_https","has_www","length_after_last_slash","nb_dots","nb_slashes","nb_hyphens","length_url","ratio_digits_url","longest_words"]]
y = df["label"]
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11430 entries, 0 to 11429
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   has_https                11430 non-null  int64  
 1   has_www                  11430 non-null  int64  
 2   length_after_last_slash  11430 non-null  float64
 3   nb_dots                  11430 non-null  int64  
 4   nb_slashes               11430 non-null  int64  
 5   nb_hyphens               11430 non-null  int64  
 6   length_url               11430 non-null  float64
 7   ratio_digits_url         11430 non-null  float64
 8   longest_words            11430 non-null  float64
dtypes: float64(4), int64(5)
memory usage: 803.8 KB


In [19]:
X.head(5)

has_https  has_www  length_after_last_slash  nb_dots  nb_slashes  \
0          0        1                 2.397895        3           3   
1          0        0                 3.496508        1           5   
2          1        0                 3.891820        4           5   
3          0        0                 2.484907        2           2   
4          0        1                 3.218876        2           5   

   nb_hyphens  length_url  ratio_digits_url  longest_words  
0           0    3.637586          0.000000       2.484907  
1           0    4.356709          0.220779       2.995732  
2           1    4.844187          0.150794       2.639057  
3           0    2.944439          0.000000       1.791759  
4           2    4.025352          0.000000       2.484907

In [20]:
# ----------------------------
# Train/Test Split (stratified)
# ----------------------------
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)
# ----------------------------
# Scaling for Logistic Regression
# ----------------------------
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [23]:
# ----------------------------
# Models
# ----------------------------
def evaluate_models(X_train_scaled, X_test_scaled, X_train_raw, X_test_raw, y_train, y_test):
    # Logistic Regression
    lr = LogisticRegression(max_iter=1000)
    lr.fit(X_train_scaled, y_train)
    y_pred_lr = lr.predict(X_test_scaled)

    # Gaussian NB
    gnb = GaussianNB()
    gnb.fit(X_train_raw, y_train)
    y_pred_gnb = gnb.predict(X_test_raw)

    # Multinomial NB (not recommended)
    mnb = MultinomialNB()
    mnb.fit(X_train_raw, y_train)
    y_pred_mnb = mnb.predict(X_test_raw)

    # Evaluation
    def eval_model(name, y_true, y_pred):
        return {
            "Model": name,
            "Accuracy": round(accuracy_score(y_true, y_pred),4),
            "Precision": round(precision_score(y_true, y_pred),4),
            "Recall": round(recall_score(y_true, y_pred),4),
            "F1 Score": round(f1_score(y_true, y_pred),4)
        }

    results = []
    results.append(eval_model("Logistic Regression", y_test, y_pred_lr))
    results.append(eval_model("Gaussian NB", y_test, y_pred_gnb))
    results.append(eval_model("Multinomial NB", y_test, y_pred_mnb))

    return pd.DataFrame(results)

# Evaluate no-log features
results = evaluate_models(
    X_train_scaled, X_test_scaled,
    X_train, X_test,
    y_train, y_test
)
print("\n=== Results ===")
print(results.to_string(index=False))


=== Results ===
              Model  Accuracy  Precision  Recall  F1 Score
Logistic Regression    0.7848     0.7747  0.8031    0.7887
        Gaussian NB    0.7463     0.8285  0.6212    0.7100
     Multinomial NB    0.7498     0.7260  0.8023    0.7623


<p style="direction: rtl; text-align: right; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">
✍️ <b>پاسخ تشریحی بخش چهارم:</b><br>
{{پاسخ_خود_را_اینجا_بنویسید}}
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش پنجم: جمع‌بندی نهایی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نتیجه‌گیری نهایی خودتان از این سوال را ارائه دهید. برای مثال باید "حداقل" به سوال‌های زیر پاسخ دهید:
<br>
- کدام مدل برای این نوع داده بهتر عمل کرده است؟ چرا؟
<br>
- کدام نوع ویژگی بیشترین تأثیر را داشته است؟ کدام نوع کمترین؟ چرا؟
<br>
- نرمال‌سازی نیاز بوده است؟ اگر بله، برای هر دو مدل؟ چرا؟
<br>
- غیر از استخراح ویژگی از خود آدرس اینترنتی، چه رویکرد دیگری می‌توان اتخاذ نمود؟
</p>


<p style="direction: rtl; text-align: right; background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111">
✍️ <b>پاسخ تشریحی بخش پنجم:</b><br>
{{پاسخ_خود_را_اینجا_بنویسید}}
</p>


# <h1 style="text-align: right;">**نکات مهم و قوانین تحویل**</h1>



<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>NLP_CA{n}_{LASTNAME}_{STUDENTID}.ipynb</code></h4>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
  <li>سلول‌های کد با برچسب <code>WRITE YOUR CODE HERE</code> را تکمیل کنید.</li>
  <li>برای پاسخ‌های متنی، متن <code>{{پاسخ_خود_را_اینجا_بنویسید}}</code> را با پاسخ خود جایگزین کنید.</li>
</ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>

 <li>
لطفاً تمامی پاسخ‌های متنی خود را با <b>فونت وزیر (Vazir)</b> و به‌صورت <b>راست‌چین</b> بنویسید.  
از استفاده از فونت‌های پیش‌فرض خودداری کنید تا ظاهر نوت‌بوک شما یک‌دست و خوانا باشد.  
در بخش‌های تشریحی، سعی کنید پاسخ‌ها را کامل، منسجم و با رعایت نگارش فارسی بنویسید.  
همچنین، به چینش تمیز سلول‌ها و اجرای درست کدها توجه کنید تا تمرین شما با فرمت خواسته‌شده و استاندارد ارائه شود.
</li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت Markdown می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>
 </ul>
    

 </div>
